File: levanta/ios/Levanta/Nucleo/Esqueleto.swift
import Foundation

/// Os treze marcos do corpo que o contador usa.
///
/// É a interseção do que o Vision (iOS) e o ML Kit Pose (Android) sabem
/// produzir, para que o mesmo motor de contagem rode idêntico nas duas
/// plataformas. A ordem e os nomes espelham `Marco.kt` de propósito: quando um
/// limiar muda de um lado, o outro tem o mesmo arquivo para mudar.
enum Marco: CaseIterable {
    case nariz
    case ombroEsq, ombroDir
    case cotoveloEsq, cotoveloDir
    case pulsoEsq, pulsoDir
    case quadrilEsq, quadrilDir
    case joelhoEsq, joelhoDir
    case tornozeloEsq, tornozeloDir
}

enum Lado { case esquerdo, direito }

/// Um ponto do corpo em coordenadas já corrigidas pela proporção da imagem.
///
/// A origem é o canto superior esquerdo e `y` cresce para baixo — convenção do
/// ML Kit. O Vision entrega o oposto (origem embaixo, `y` para cima), então o
/// adaptador da câmera inverte o eixo antes de construir o esqueleto.
struct Ponto {
    let x: Float
    let y: Float
    let confianca: Float
}

/// Um quadro de pose: os marcos detectados e o instante em que foram vistos.
struct Esqueleto {
    /// Abaixo disto o marco é tratado como não visto.
    static let confiancaMinima: Float = 0.35

    private let marcos: [Marco: Ponto]
    let instanteMs: Int64

    /// Monta um esqueleto a partir de marcos normalizados em 0...1, corrigindo
    /// `x` pela proporção (largura ÷ altura da imagem).
    ///
    /// Sem essa correção a imagem fica achatada num quadrado e todo ângulo sai
    /// distorcido: o mesmo agachamento mediria valores diferentes conforme o
    /// celular estivesse em pé ou deitado.
    init(marcos: [Marco: Ponto], proporcao: Float, instanteMs: Int64) {
        let p = proporcao > 0 ? proporcao : 1
        self.marcos = marcos.mapValues { Ponto(x: $0.x * p, y: $0.y, confianca: $0.confianca) }
        self.instanteMs = instanteMs
    }

    /// O ponto, ou `nil` se ausente ou pouco confiável.
    func ponto(_ m: Marco) -> Ponto? {
        guard let p = marcos[m], p.confianca >= Esqueleto.confiancaMinima else { return nil }
        return p
    }

    func temTodos(_ ms: Marco...) -> Bool { ms.allSatisfy { ponto($0) != nil } }
}

/// Ângulo em graus no vértice `b`, entre 0 e 180. `nil` se algum ponto faltar.
func angulo(_ a: Ponto?, _ b: Ponto?, _ c: Ponto?) -> Float? {
    guard let a, let b, let c else { return nil }
    let abx = a.x - b.x, aby = a.y - b.y
    let cbx = c.x - b.x, cby = c.y - b.y
    let normas = (abx * abx + aby * aby).squareRoot() * (cbx * cbx + cby * cby).squareRoot()
    guard normas > 1e-6 else { return nil }
    let cosseno = max(-1, min(1, (abx * cbx + aby * cby) / normas))
    return acos(cosseno) * 180 / .pi
}

/// Distância euclidiana entre dois pontos. `nil` se algum faltar.
func distancia(_ a: Ponto?, _ b: Ponto?) -> Float? {
    guard let a, let b else { return nil }
    let dx = a.x - b.x, dy = a.y - b.y
    return (dx * dx + dy * dy).squareRoot()
}

/// Inclinação do segmento a→b em relação à horizontal, de 0 a 90 graus.
/// Perto de 0 o tronco está deitado; perto de 90, em pé.
func inclinacao(_ a: Ponto?, _ b: Ponto?) -> Float? {
    guard let a, let b else { return nil }
    let dx = abs(a.x - b.x), dy = abs(a.y - b.y)
    guard dx > 1e-6 || dy > 1e-6 else { return nil }
    return atan2(dy, dx) * 180 / .pi
}

/// Ponto médio entre dois pontos, com a menor das duas confianças.
func medio(_ a: Ponto?, _ b: Ponto?) -> Ponto? {
    guard let a, let b else { return nil }
    return Ponto(x: (a.x + b.x) / 2, y: (a.y + b.y) / 2, confianca: min(a.confianca, b.confianca))
}

/// Reescala `v` do intervalo [de, ate] para 0...1, saturando fora dele.
func faixa(_ v: Float, _ de: Float, _ ate: Float) -> Float {
    guard abs(ate - de) > 1e-6 else { return 0 }
    return max(0, min(1, (v - de) / (ate - de)))
}

File: levanta/android/core/src/main/kotlin/br/com/levanta/core/pose/Esqueleto.kt
package br.com.levanta.core.pose

import kotlin.math.abs
import kotlin.math.acos
import kotlin.math.hypot
import kotlin.math.min

/**
 * Os treze marcos do corpo que o contador usa.
 *
 * É a interseção do que o ML Kit Pose (33 marcos, Android) e o Vision
 * (19 juntas, iOS) sabem produzir, para que o mesmo motor de contagem
 * rode idêntico nas duas plataformas.
 */
enum class Marco {
    NARIZ,
    OMBRO_ESQ, OMBRO_DIR,
    COTOVELO_ESQ, COTOVELO_DIR,
    PULSO_ESQ, PULSO_DIR,
    QUADRIL_ESQ, QUADRIL_DIR,
    JOELHO_ESQ, JOELHO_DIR,
    TORNOZELO_ESQ, TORNOZELO_DIR,
}

/** Lado do corpo. O contador escolhe o mais visível a cada quadro. */
enum class Lado { ESQUERDO, DIREITO }

/**
 * Um ponto do corpo em coordenadas já corrigidas pela proporção da imagem.
 *
 * `x` e `y` saem normalizados em 0..1 pela plataforma, o que achata a imagem
 * num quadrado e distorce todo ângulo medido nela. [Esqueleto.de] devolve o `x`
 * multiplicado pela proporção largura/altura, restaurando a escala real — sem
 * isso um agachamento filmado em retrato mede um ângulo de joelho diferente do
 * mesmo agachamento filmado em paisagem.
 *
 * A origem é o canto superior esquerdo e `y` cresce para baixo, convenção do
 * ML Kit; o adaptador do iOS inverte o eixo do Vision antes de chegar aqui.
 */
data class Ponto(val x: Float, val y: Float, val confianca: Float)

/** Um quadro de pose: os marcos detectados e o instante em que foram vistos. */
class Esqueleto private constructor(
    private val marcos: Map<Marco, Ponto>,
    val instanteMs: Long,
) {
    companion object {
        /** Abaixo disto o marco é tratado como não visto. */
        const val CONFIANCA_MINIMA = 0.35f

        /**
         * Monta um esqueleto a partir de marcos normalizados em 0..1,
         * corrigindo `x` pela [proporcao] (largura ÷ altura da imagem).
         */
        fun de(marcos: Map<Marco, Ponto>, proporcao: Float, instanteMs: Long): Esqueleto {
            val p = if (proporcao > 0f) proporcao else 1f
            return Esqueleto(marcos.mapValues { (_, v) -> v.copy(x = v.x * p) }, instanteMs)
        }
    }

    /** O ponto, ou `null` se ausente ou pouco confiável. */
    fun ponto(m: Marco): Ponto? = marcos[m]?.takeIf { it.confianca >= CONFIANCA_MINIMA }

    /** Todos os marcos existem e são confiáveis? */
    fun temTodos(vararg ms: Marco): Boolean = ms.all { ponto(it) != null }

    /** Confiança média dos marcos pedidos; 0 se algum faltar. */
    fun confiancaMedia(vararg ms: Marco): Float {
        var soma = 0f
        for (m in ms) soma += (ponto(m) ?: return 0f).confianca
        return soma / ms.size
    }
}

/** Ângulo em graus no vértice [b], entre 0 e 180. `null` se algum ponto faltar. */
fun angulo(a: Ponto?, b: Ponto?, c: Ponto?): Float? {
    if (a == null || b == null || c == null) return null
    val abx = a.x - b.x; val aby = a.y - b.y
    val cbx = c.x - b.x; val cby = c.y - b.y
    val normas = hypot(abx, aby) * hypot(cbx, cby)
    if (normas < 1e-6f) return null
    val cos = ((abx * cbx + aby * cby) / normas).coerceIn(-1f, 1f)
    return Math.toDegrees(acos(cos).toDouble()).toFloat()
}

/** Distância euclidiana entre dois pontos. `null` se algum faltar. */
fun distancia(a: Ponto?, b: Ponto?): Float? {
    if (a == null || b == null) return null
    return hypot(a.x - b.x, a.y - b.y)
}

/**
 * Inclinação do segmento a→b em relação à horizontal, de 0 a 90 graus.
 * Serve para saber se o tronco está deitado (perto de 0) ou em pé (perto de 90).
 */
fun inclinacao(a: Ponto?, b: Ponto?): Float? {
    if (a == null || b == null) return null
    val dx = abs(a.x - b.x); val dy = abs(a.y - b.y)
    if (dx < 1e-6f && dy < 1e-6f) return null
    return Math.toDegrees(kotlin.math.atan2(dy, dx).toDouble()).toFloat()
}

/** Ponto médio entre dois pontos, com a menor das duas confianças. */
fun medio(a: Ponto?, b: Ponto?): Ponto? {
    if (a == null || b == null) return null
    return Ponto((a.x + b.x) / 2f, (a.y + b.y) / 2f, min(a.confianca, b.confianca))
}

/** Reescala [v] do intervalo [de, ate] para 0..1, saturando fora dele. */
fun faixa(v: Float, de: Float, ate: Float): Float {
    if (abs(ate - de) < 1e-6f) return 0f
    return ((v - de) / (ate - de)).coerceIn(0f, 1f)
}
